### PR TITLE
Fix: Filter ContentExpiry changelist results by site

### PR DESCRIPTION
**Pull request by Aiky30**
This Pull request adds the ability for content_expiry and third party packages to pre filter the ContentExpiry Change list queryset. The following decisions have been made based on the justifications included within:

- The django-cms PageContent model has a builtin filter because it is a high order package that content expiry is dependant on.

- The djangocms_alias implementation to filter any objects by site is based on the fact that djangocms_alias is a core high level package, to correctly "add" features to another package, the test suite would need to ensure that the functionality continues to work and exist and therefore becomes "dependant". Lower level packages that wish to be "content expiry compatible" should register any functionality in the packages cms_config and "depend" on content-expiry.

The query solution has 2 options to filter the content objects. There are two trains of thought here:

1. Blindly find all Content objects with site that aren't for the current site and exclude them from the Content Expiry query, this gives 1 huge query for each content type to exclude that only gets larger the longer the site receives changes. 

```python
        # PageContent = Expiry->Version->Content->Page->Node->Site
        from cms.models import PageContent

        def get_page_content_site_objects(site):
            page_queryset = PageContent._original_manager.exclude(page__node__site=site)
            return page_queryset.select_related('page__node')

        # Get all content types for site
        # https://docs.djangoproject.com/en/3.2/ref/contrib/contenttypes/#reverse-generic-relations
        page_content_ctype = ContentType.objects.get_for_model(PageContent)
        exclusion_list = get_page_content_site_objects(current_site)
        queryset = queryset.exclude(
            version__content_type=page_content_ctype, version__object_id__in=exclusion_list
        )
```

2. The alternative is to use the Generic relationship in reverse and exclude on the items in the fetched set that have the site. This excludes only the items in the original queryset. In order to keep this low impact the queries to filter on content from sites would need to happen after the user has selected the user controllable filters, this is because the users filters limit the result set and without them fetches all of the data for all ContentExpiry records which would perform worse than option 1.

```python
        # PageContent = Expiry->Version->Content->Page->Node->Site
        from django.db import models
        from cms.models import PageContent

        def filter_expiry_page_content_objects(site, queryset):
            """
            Returns a queryset list of ContentExpiry records that are attached to a page
            and are not part of the site supplied and should be excluded!
            """
            page_content_ctype = ContentType.objects.get_for_model(PageContent)

            # Generic foreign key can't currently handle exclusions on a generic foreign key:
            # queryset.exclude(version__cms_pagecontent__page__node__site=site)
            # https://code.djangoproject.com/ticket/26261
            expiry_othersite_queryset = queryset.annotate(
                is_bound_to_site=models.Case(
                    models.When(
                        Q(version__content_type=page_content_ctype) |
                        ~Q(version__cms_pagecontent__page__node__site=site),
                        then=models.Value(True),
                    ),
                    default=models.Value(False),
                    output_field=models.BooleanField(),
                )
            ).filter(is_bound_to_site=False)

            return expiry_othersite_queryset


        # For each content type with site limit the queryset
        current_site = get_current_site(request)
        exclusion_list = filter_expiry_page_content_objects(current_site, queryset)
        queryset = queryset.exclude(pk__in=exclusion_list.values('pk'))
```

Both methods work and have their own inefficiencies, one thing that option 2 can give is the control to ensure that the query burden is scoped to the filter controls which is not the case for option 1 where the exclusion set is always a huge list of everything not available on the current site. However, this is not the case when the queryset is seen before any filters are applied which is the order of the chnagelist where the get_queryset() method is hit first and then the filters are applied, this would result in option 2 performing an exclusion the same as option 1 in a much less efficient approach!

The implementation should be flexible enough for  any 3rd party developers to implement and choose which version to use. As this package requires the CMS we have the pleasure of deciding how the PageContent site filtering should work. :-)

### CODE REVIEW

**Review by Bernardvdv — APPROVED** [review]
Big change :-). Just need an entry in the changelog but looks good to me

File: djangocms_content_expiry/conf.py [hunk 1]
```diff
@@ -15,3 +15,8 @@
 DEFAULT_CONTENT_EXPIRY_EXPORT_DATE_FORMAT = getattr(
     settings, "CMS_CONTENT_EXPIRY_DEFAULT_CONTENT_EXPIRY_EXPORT_DATE_FORMAT", "%Y/%m/%d %H:%M %z"
 )
+
+# Default Content Expiry changelist page content exclusion cache expiration duration is seconds
```
Comment (adam-murray) [inline djangocms_content_expiry/conf.py:19]: I think this should say 'in seconds'

**Review by adam-murray — APPROVED** [review]

File: djangocms_content_expiry/conf.py (hunk 1 above)
Comment (Aiky30) [reply djangocms_content_expiry/conf.py:19]:
```suggestion
# Default Content Expiry changelist page content exclusion cache expiration duration in seconds
```

**Review by Bernardvdv — APPROVED** [review]

File: CHANGELOG.rst [hunk 2]
```diff
@@ -4,6 +4,7 @@ Changelog
 
 unreleased
 ==========
+* feat: Filter the Content Expiry Chnagelist PageContents Content Type by site, the feature can also be used by third parties to restrict the entries shown in the changelist.
```
Comment (Aiky30) [inline CHANGELOG.rst:7]:
```suggestion
* feat: Filter the Content Expiry Changelist PageContents and Alias Content Type by site, the feature can also be used by other third party packages to restrict the entries shown in the changelist.
```

**Review by Bernardvdv — APPROVED** [review]

**Review by adam-murray — APPROVED** [review]
Only noticed a small spelling error, otherwise, it's all good!

File: djangocms_content_expiry/cms_config.py [hunk 3]
```diff
@@ -1,13 +1,29 @@
...
+    get_changelist_page_content_exclusion_cache,
+    set_changelist_page_content_exclusion_cache,
+)
 from .constants import CONTENT_EXPIRY_EXPIRE_FIELD_LABEL
 
 
+try:
+    from djangocms_alias.models import AliasContent
+    djangocms_alias_enabled = True
+except (ImportError):
```
Comment (Aiky30) [inline djangocms_content_expiry/cms_config.py:23]:
```suggestion
except ImportError:
```

File: djangocms_content_expiry/admin.py [hunk 4]
```diff
@@ -42,6 +43,17 @@ class Media:
             )
         }
 
+    def get_queryset(self, request):
+        queryset = super().get_queryset(request)
+
+        # Execute any filters that needs to be added before any of the admin filters
```
[the remark this answers is not in the document]
Comment (Aiky30) [reply djangocms_content_expiry/admin.py:49]:
```suggestion
        # Execute any filters that need to be added before any of the admin filters
```

**Pull request merged**